crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1308 @@
+#![allow(missing_docs)] // FIXME
+
+#[derive(Debug)]
+pub enum Error {
+    /// There was a type reference that was canonicalized and no longer
+    /// references an index into a module's types space, so we cannot encode it
+    /// into a Wasm binary again.
+    CanonicalizedHeapTypeReference,
+    /// The const expression is invalid: not actually constant or something like
+    /// that.
+    InvalidConstExpr,
+    /// There was an error when parsing.
+    ParseError(wasmparser::BinaryReaderError),
+}
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::ParseError(_e) => {
+                write!(fmt, "There was an error when parsing")
+            }
+            Self::InvalidConstExpr => write!(fmt, "The const expression was invalid"),
+            Self::CanonicalizedHeapTypeReference => write!(
+                fmt,
+                "There was a canonicalized heap type reference without type index information"
+            ),
+        }
+    }
+}
+
+impl std::error::Error for Error {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            Self::ParseError(e) => Some(e),
+            Self::InvalidConstExpr | Self::CanonicalizedHeapTypeReference => None,
+        }
+    }
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub trait WasmParserToWasmEncoder {
+    fn core_module<'a, 'b>(
+        &mut self,
+        module: &'a mut crate::Module,
+        sections: impl Iterator<Item = Result<wasmparser::Payload<'b>, wasmparser::BinaryReaderError>>,
+    ) -> Result<&'a crate::Module> {
+        utils::core_module(self, module, sections)
+    }
+
+    fn component_primitive_val_type(
+        &mut self,
+        ty: wasmparser::PrimitiveValType,
+    ) -> crate::component::PrimitiveValType {
+        utils::component_primitive_val_type(self, ty)
+    }
+
+    fn memory_index(&mut self, memory: u32) -> u32 {
+        utils::memory_index(self, memory)
+    }
+
+    fn mem_arg(&mut self, arg: wasmparser::MemArg) -> crate::MemArg {
+        utils::mem_arg(self, arg)
+    }
+
+    fn ordering(&mut self, arg: wasmparser::Ordering) -> crate::Ordering {
+        utils::ordering(self, arg)
+    }
+
+    fn function_index(&mut self, func: u32) -> u32 {
+        utils::function_index(self, func)
+    }
+
+    fn tag_index(&mut self, tag: u32) -> u32 {
+        utils::tag_index(self, tag)
+    }
+
+    fn label_index(&mut self, label: u32) -> u32 {
+        utils::label_index(self, label)
+    }
```

Comment:
For functions like this where the `utils` version is blank I think it's ok to remove the `utils` version

Reply:
Also sorting-wise could all the indices be moved up to the top of the trait to more clearly call out that it's possible to remap them?

Reply:
I've reordered the members, though I'm unsure about the order